src/librustc/middle/ty/mod.rs

```diff
@@ -2396,6 +2396,45 @@ impl<'tcx> ctxt<'tcx> {
             || self.sess.cstore.item_super_predicates(self, did))
     }
 
+    /// If `type_needs_drop` returns true, then `ty` is definitely
+    /// non-copy and *might* have a destructor attached; if it returns
+    /// false, then `ty` definitely has no destructor (i.e. no drop glue).
+    ///
+    /// (Note that this implies that if `ty` has a destructor attached,
+    /// then `type_needs_drop` will definitely return `true` for `ty`.)
+    pub fn type_needs_drop(&self, ty: Ty<'tcx>) -> bool {
+        self.type_needs_drop_given_env(ty, &self.empty_parameter_environment())
+    }
+
+    /// Core implementation of type_needs_drop, potentially making use of
+    /// and/or updating caches held in the `param_env`.
+    pub fn type_needs_drop_given_env<'a>(&self,
+                                         ty: Ty<'tcx>,
+                                         param_env: &ty::ParameterEnvironment<'a,'tcx>) -> bool {
+        // Issue #22536: We first query type_moves_by_default.  It sees a
+        // normalized version of the type, and therefore will definitely
```

Comment:
The same note that @nagisa wrote on `fn type_needs_drop` applies here, namely, that the assumptions that hold in the context of `trans` may not hold here. (When I wrote "normalized" in the original comment, I probably should have written "monomorphized.")